### PR TITLE
zephyr: Use upstream core startup API

### PR DESCRIPTION
**Pull request by andyross**
The original Zephyr version of arch_cpu_enable_core() duplicated logic
internal to the Zephyr SMP layer.  This is now exposed as an internal
API that is more easily testable upstream.  Use that instead.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

### CODE REVIEW

**Comment by andyross** [comment]
Submitted as example code, really.  I don't have the ability to run true multicore SOF tests on my hardware, I can only get as far as initialization (which seems to work for me fine on TGL).  But this is the "supported" Zephyr runtime smp startup mechanism, and as of today there's an upstream test in review for it, at least.

**Comment by andyross** [comment]
Sorry, I think I had mentioned this verbally or via chat, but to call it out explicitly:

This patch removes a call to pm_runtime_get() in the original code.  I don't know if that's correct or not, honestly.  My brief read is that this code is entirely driver-level in SOF and responsible only for hardware stuff the Zephyr code is now doing.  If that's not correct, it's surely going to break something.

Also: note that this currently needs https://github.com/zephyrproject-rtos/zephyr/pull/38319 applied to Zephyr because of a build glitch (though we could fix and merge that separately from the cleanup PR).

File: zephyr/wrapper.c [hunk 1]
```diff
@@ -703,37 +703,12 @@ void platform_dai_wallclock(struct comp_dev *dai, uint64_t *wallclock)
...
-
-	/* Power up secondary core */
-	pm_runtime_get(PM_RUNTIME_DSP, PWRD_BY_TPLG | id);
-
-	arch_start_cpu(id, z_interrupt_stacks[id], CONFIG_ISR_STACK_SIZE,
-		       secondary_init, &start_flag);
+	extern void z_smp_start_cpu(int id);
 
-	atomic_set(&start_flag, 1);
+	z_smp_start_cpu(id);
```
Comment (kv2019i) [inline zephyr/wrapper.c:711]: The pm_runtime_get() is a bit worrisome, but I went (again) through the actions on e.g. sof/src/platform/intel/cavs/lib/pm_runtime.c and it does seem that it's all replaced by Zephyr core management.

One thing I noticed is that src/platform/intel/cavs/platform.c:platform_boot_complete() is powering up the last core on TGL, and uses this pm_runtime_get() to do it (with bit PWRD_BY_HPRO). I think to preserve these on Zephyr, we'd need to provide wrapping from pm_runtime_get() -> calls to z_smp_start_cpu(). But that's not really business of this PR.

**Review by kv2019i — COMMENTED** [review]
Build error still, but otherwise looks good.

**Comment by andyross** [comment]
FWIW: I split the Zephyr build fix out separately as a single patch in https://github.com/zephyrproject-rtos/zephyr/pull/38368 , which might have a chance of merging during the current release cycle (it's definitely a bug fix).

**Review by lgirdwood — COMMENTED** [review]
Seeing code removal in the wrapper is good to see (as the wrapper will one day completely evaporate).
@kv2019i re the PM call, I assume this is for powering On/OFF cores and we are using this elsewhere in SOF today (using ref count) ? 
@andyross we can get this in for v1.9, but will need to get our Zephyr CI back up and building before we merge.

**Comment by andyross** [comment]
FWIW: the Zephyr build fix merged, so this should be able to build successfully against current main.  I don't know how to kick the CI in SOF though.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
> FWIW: the Zephyr build fix merged, so this should be able to build successfully against current main. I don't know how to kick the CI in SOF though.

With the command above :)

**Comment by lgirdwood** [comment]
@andyross this may need rebased on HEAD  and forced pushed to pick up the Zephyr fixed that's now merged.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Hmm, I was testing your adsp cleanup patch to Zephyr git, but kept getting failures with multitcore TPLG on TGL. It turns out it was this patch that causes the the CORE_ENABLE IPC to fail.

Not sure what's happening, but I'll log a minus so this doesn't get merged before we understsand the failure.

**Comment by andyross** [comment]
Seems not unlikely that the SOF pm call is doing something to en/disable/whatever the IPC interrupt?  Or scheduling/not-scheduling a task that changes CPU assignments?   This is sort of what I was worried about.

Would be worth trying to add just that one line back, though I'd be afraid of other interactions given it's duplicating a lot of hardware interfacing that Zephyr is now doing.

Note that Zephyr has no IPC driver right now, so all masking and management is happening in SOF exclusively.  We should probably migrate that driver first...

**Comment by lgirdwood** [comment]
> Note that Zephyr has no IPC driver right now, so all masking and management is happening in SOF exclusively. We should probably migrate that driver first...

Zephyr has the CAVS v1.5 IPC IIRC, I think I've discussed with @dcpleung in the past. We need to keep the current v1.5 code and add in the differences for v1.8 - v2.5 (same IP). I dont think it's too big a job since the interface is in place already.

**Comment by andyross** [comment]
Ancient.  Close.

**Pull request closed without merging**